### DISCUSSION TITLE
Coroutine Context in Custom Transition for 2.0

### DISCUSSION THREAD

**Discussion by tyczj** [coil-kt/coil#1196] [Q&A] [answered]
I am trying to upgrade from 1.4 to 2.0 and I see the transition usage has changed. The documentation points to the [CrossfadeTransition](https://github.com/coil-kt/coil/blob/main/coil-base/src/main/java/coil/transition/CrossfadeTransition.kt) which is fine but in 1.4 you could mark the `transition` method as suspend to do some coroutine work to create the custom transition in it but it looks like in 2.0 you cant do this anymore.

How would you do coroutine work in the transition method in 2.0?

this is what I was doing in 1.4

```
override suspend fun transition(target: TransitionTarget, result: ImageResult) {

        // Compute the palette on a background thread.
        if (result is SuccessResult) {
            val bitmap = result.drawable.toBitmap()
            val palette = withContext(Dispatchers.IO) {
                Palette.Builder(bitmap).generate()
            }
            onGenerated(palette)
        }

        // Execute the delegate transition.
        val delegateJob = delegate?.let { delegate ->
            coroutineScope {
                launch(Dispatchers.Main.immediate) {
                    delegate.transition(target, result)
                }
            }
        }

        delegateJob?.join()
    }
```

**Comment by colinrtwhite** [comment, marked as answer]
I'd recommend using a `listener` in 2.0 to generate the palette. [Check out this snippit for how to do that](https://coil-kt.github.io/coil/recipes/#palette). Transitions are no longer suspend functions in 2.x as it's not necessary to suspend until the transition is complete; this was required in 1.x for bitmap pooling.
